go/arrow/datatype_nested.go

```diff
@@ -384,6 +388,286 @@ func (t *MapType) Layout() DataTypeLayout {
+		}
+	}
+	return nil
+}
+
+func (t *unionType) MaxTypeCode() (max UnionTypeCode) {
+	if len(t.typeCodes) == 0 {
+		return
+	}
+
+	max = t.typeCodes[0]
+	for _, c := range t.typeCodes[1:] {
+		if c > max {
+			max = c
+		}
+	}
+	return
+}
+
+func (t *unionType) String() string {
+	var b strings.Builder
+	b.WriteByte('<')
+	for i := range t.typeCodes {
+		if i != 0 {
+			b.WriteString(", ")
+		}
+		fmt.Fprintf(&b, "%s=%d", t.children[i], t.typeCodes[i])
+	}
+	b.WriteByte('>')
+	return b.String()
+}
+
+func (t *unionType) fingerprint() string {
+	var b strings.Builder
+	for _, c := range t.typeCodes {
+		fmt.Fprintf(&b, ":%d", c)
+	}
+	b.WriteString("]{")
+	for _, c := range t.children {
+		fingerprint := c.Fingerprint()
+		if len(fingerprint) == 0 {
+			return ""
+		}
+		b.WriteString(fingerprint)
+		b.WriteByte(';')
+	}
+	b.WriteByte('}')
+	return b.String()
+}
+
+func fieldsFromArrays(arrays []Array, names ...string) (ret []Field) {
+	ret = make([]Field, len(arrays))
+	if len(names) == 0 {
+		for i, c := range arrays {
+			ret[i] = Field{Name: strconv.Itoa(i), Type: c.DataType(), Nullable: true}
+		}
+	} else {
+		debug.Assert(len(names) == len(arrays), "mismatch of arrays and names")
+		for i, c := range arrays {
+			ret[i] = Field{Name: names[i], Type: c.DataType(), Nullable: true}
+		}
+	}
+	return
+}
+
+// SparseUnionType is the concrete type for Sparse union data.
+//
+// A sparse union is a nested type where each logical value is taken
+// from a single child. A buffer of 8-bit type ids indicates which child
+// a given logical value is to be taken from.
+//
+// In a sparse union, each child array will have the same length as the
+// union array itself, regardless of the actual number of union values which
+// refer to it.
+//
+// Unlike most other types, unions do not have a top-level validity bitmap.
+type SparseUnionType struct {
+	unionType
+}
+
+// SparseUnionFromArrays enables creating a union type from a list of Arrays,
+// field names, and type codes. len(fields) should be either 0 or equal to len(children).
+// len(codes) should also be either 0, or equal to len(children).
+//
+// If len(fields) == 0, then the fields will be named numerically as "0", "1", "2"...
+// and so on. If len(codes) == 0, then the type codes will be constructed as
+// [0, 1, 2, ..., n].
+func SparseUnionFromArrays(children []Array, fields []string, codes []UnionTypeCode) *SparseUnionType {
+	if len(codes) == 0 {
+		codes = make([]UnionTypeCode, len(children))
+		for i := range children {
+			codes[i] = UnionTypeCode(i)
+		}
+	}
+	return SparseUnionOf(fieldsFromArrays(children, fields...), codes)
+}
+
+// SparseUnionOf is equivalent to UnionOf(arrow.SparseMode, fields, typeCodes),
+// constructing a SparseUnionType from a list of fields and type codes.
+//
+// If len(fields) != len(typeCodes) this will panic. They are allowed to be
+// of length 0.
+func SparseUnionOf(fields []Field, typeCodes []UnionTypeCode) *SparseUnionType {
+	ret := &SparseUnionType{}
```

Comment:
Again for consistency use new(SparseUnionType )

Reply:
so, it would actually be more consistent to swap out the usages of `new` since instantiating using `&Type{}` is the consistent style in the existing codebase. (look elsewhere at the other `*Of(...)` functions or any of the `New*Data` functions in the array package). I just didn't change the usages of `new(strings.Builder)` since I wasn't modifying the existing code. I'll swap them out here in order to make that more consistent.